### ISSUE TITLE
DeferredLightingRenderer breaks layer depth

### ISSUE THREAD

**Issue by RobinPerris** [prime31/Nez#42]
Currently using the following system to change the layer depth of objects based on their position on the screen:

```csharp
public class RenderOrderSystem : PassiveSystem
{
    protected override void begin()
    {
        var sprites = scene.renderableComponents.componentsWithRenderLayer((int)Layers.Object);

        if (sprites.Count == 0)
            return;

        var max = 1.0f / sprites.Count;
        var index = 0;

        foreach (var sprite in sprites.OrderByDescending(s => s.entity.transform.position.Y))
        {
            index++;
            var depth = max * index;
            sprite.setLayerDepth(depth);
        }
    }
}
```

and the following code to set up the lighting renderer:

```csharp
var lightingRenderer = new DeferredLightingRenderer(0, (int)Layers.Light);
lightingRenderer.renderLayers = new int[] { (int)Layers.Ground, (int)Layers.Object, (int)Layers.Fringe };
```

It works fine normally:
![No lighting - okay](http://i.imgur.com/zp7edQy.png)

But breaks when a `DeferredLightingRenderer` is added:
![Lighting - broken](http://i.imgur.com/k7Due1m.png)

Am I missing any steps required to get this to work properly?

**Comment by prime31** [comment]
The DeferredLightingRenderer (and any renderers with a renderLayers field) have two modes of operation:

- if the renderLayers are passed into the constructor they are sorted for you
- if you set the renderLayers field the order you set will be used

If I had go guess, it would be this that is tricking you. Try just passing your renderLayers into the constructor and let the DeferredLightingRenderer sort them for you:

`new DeferredLightingRenderer( 0, (int)Layers.Light, (int)Layers.Ground, (int)Layers.Object, (int)Layers.Fringe )`

**Comment by RobinPerris** [comment]
Render layers work fine, it's a problem with the renderable component lists not being re-sorted after changing the layer depth.

This works fine in the `DefaultRenderer` but doesn't in `DeferredLightingRenderer`.

Currently the default renderer is directly referencing the array of renderable components whilst the deferred lighting renderer is pulling through a pre-filtered list of components using the `componentsWithRenderLayer` method in the `RenderableComponentList` class.

Looking in to it the list which is passed through from this method is _not_ re-sorted when the layer depth of components is changed - it's only re-sorted after an entirely new component is added to the collection.

This explains the weird issues I saw in my game - if I dropped an item, all of the map objects would suddenly be in the right order.

Render layers need to be added to the `_unsortedRenderLayers` list to queue them up to be re-sorted efficiently. I believe the code needs to be modified to add render layers to this list when renderable components within that render layer have their layer depth updated.

I think the correct way to fix this would be to change the `setLayerDepth` method in `RenderableComponent` to specify which render layer needs to be updated.

See pull request #43
